### PR TITLE
[RLlib] Layout of Trajectory View API (new class: Trajectory; not used yet).

### PR DESCRIPTION
**Pull request by sven1977**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/evaluation/trajectory.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                                   values: Dict[str, TensorType]) -> None:
+        """Add the given dictionary (row) of values to this batch.
+
+        Args:
+            env_id (EnvID): Unique id for the episode we are adding the initial
+                observation for.
+            agent_id (AgentID): Unique id for the agent we are adding the
+                initial observation for.
+            policy_id (PolicyID): Unique id for policy controlling the agent.
+            values (Dict[str, TensorType]): Data dict (interpreted as a single
```
Comment (ericl) [inline rllib/evaluation/trajectory.py:123]: Why not pass these separately?

File: rllib/models/modelv2.py [hunk 2]
```diff
@@ -266,6 +266,10 @@ def get_view_requirements(
         # Single requirement: Pass current obs as input.
         return {
             SampleBatch.CUR_OBS: ViewRequirement(timesteps=0),
+            SampleBatch.PREV_ACTIONS:
+                ViewRequirement(SampleBatch.ACTIONS, timesteps=-1),
```
Comment (ericl) [inline rllib/models/modelv2.py:270]: Why -1 instead of 0?

File: rllib/evaluation/trajectory.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+from ray.rllib.utils.numpy import convert_to_numpy
+from ray.rllib.utils.types import AgentID, EnvID, PolicyID, TensorType
+
+tf1, tf, tfv = try_import_tf()
+torch, _ = try_import_torch()
+
+logger = logging.getLogger(__name__)
+
+
+class Trajectory:
```
Comment (ericl) [inline rllib/evaluation/trajectory.py:16]: Can we have unit tests for `Trajectory` class covering the edge cases here?

**Review by ericl — CHANGES_REQUESTED** [review]

File: rllib/evaluation/trajectory.py (hunk 3 above)
Comment (sven1977) [reply rllib/evaluation/trajectory.py:16]: sure ...

File: rllib/models/modelv2.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/modelv2.py:270]: b/c it's the prev-action:
t=0: add obs(0) -> prev-action="np.zeros"-action
t=1: add a(0), r(0), obs(1) -> prev-action=a(t-1=0)
t=2: add a(1), r(1), obs(2) -> prev-action=a(t-1=1)

File: rllib/evaluation/trajectory.py (hunk 1 above)
Comment (sven1977) [reply rllib/evaluation/trajectory.py:123]: As **kwargs? It's so much more slower, if I remember correctly:
The call to `add_values/add_action_reward_next_obs` is the most expensive one in the whole sampler loop, by far.

```
import time

num = 1000000

def f(d):
    return

def f_kwargs(**d):
    return

d_in = {str(i): 0 for i in range(100)}
start = time.time_ns()
for _ in range(num):
    f(d_in)
delta = time.time_ns() - start
print("delta / num={}".format(delta / num))

start = time.time_ns()
for _ in range(num):
    f_kwargs(**d_in)
delta = time.time_ns() - start
print("delta / num={}".format(delta / num))
```

Comment (sven1977) [reply rllib/evaluation/trajectory.py:123]: Yup, output:
```
delta / num=124.0005
delta / num=3793.9991
```

File: rllib/evaluation/trajectory.py (hunk 3 above)
Comment (sven1977) [reply rllib/evaluation/trajectory.py:25]: This was added. Note that there will be more performance optimizations coming up over the next PRs. Will add a performance test (new sampling/trajectory view API vs current implementation(s)) as well.

**Review by ericl — APPROVED** [review]

**Pull request merged**